File: Sources/macOS/Sources/OnScreen/Features/FantozziCloud.swift
import Foundation
import Yage

extension ScreenEnvironment {
    func scheduleRainyCloud() {
        scheduleRandomly(withinHours: 0..<5) { [weak self] in
            guard let self else { return }
            guard self.settings.randomEvents else { return }
            guard let victim = self.randomPet() else { return }
            self.rainyCloudUseCase.start(target: victim, world: self)
        }
    }
}

protocol RainyCloudUseCase {
    func start(target: Entity, world: World)
}

class RainyCloudUseCaseImpl: RainyCloudUseCase {
    func start(target: Entity, world: World) {
        let cloud = buildCloud(at: target.frame.origin, in: world)
        setupSeeker(for: cloud, to: target)
        scheduleCleanUp(cloud: cloud, world: world)
    }
    
    private func buildCloud(at origin: CGPoint, in world: World) -> Entity {
        let cloud = PetEntity(of: .cloud, in: world)
        cloud.frame.size = CGSize(
            width: cloud.frame.size.width * 2,
            height: cloud.frame.size.height * 2
        )
        cloud.frame.origin = origin
        cloud.isEphemeral = true
        world.children.append(cloud)
        return cloud
    }
    
    private func setupSeeker(for cloud: Entity, to target: Entity) {
        let yOffset = cloud.frame.height - target.frame.height
        let seeker = Seeker()
        cloud.install(seeker)
        seeker.follow(
            target,
            to: .above,
            offset: CGSize(width: 0, height: yOffset),
            autoAdjustSpeed: true
        ) { _ in }
    }
    
    private func scheduleCleanUp(cloud: Entity, world: World) {
        let duracy = TimeInterval.random(in: 60..<120)
        DispatchQueue.main.asyncAfter(deadline: .now() + duracy) { [weak world] in
            cloud.kill()
            world?.children.remove(cloud)
        }
    }
}

private extension Species {
    static let cloud = Species(
        id: "fantozzi",
        capabilities: [
            "AnimatedSprite",
            "AnimationsProvider",
            "LinearMovement",
            "PetsSpritesProvider"
        ],
        dragPath: "front",
        movementPath: "front",
        speed: 2,
        zIndex: 200
    )
}

File: Sources/windows/onscreen/features/fantozzi_cloud.py
from random import randint
from PyQt6.QtCore import QTimer
from onscreen.models.pet_entity import PetEntity
from yage.capabilities.seeker import Seeker, SeekerTargetPosition
from yage.models.entity import Entity
from yage.models.species import Species
from yage.models.world import World
from yage.utils.geometry import Point


class RainyCloudUseCase:
    def __init__(self):
        self.target = None
        self.cloud = None
        self._completion_date = None

    def start(self, target: Entity, world: World):
        self.target = target
        self.cloud = self._build_cloud(target.frame.top_left, world)
        self._setup_seeker()
        self._schedule_completion()

    def _schedule_completion(self):
        duration = randint(60, 120) * 1000
        QTimer.singleShot(duration, lambda: self._cleanup())

    def _build_cloud(self, origin: Point, world: World) -> Entity:
        cloud = PetEntity(self._cloud_species(), world)
        cloud.frame.set_size(cloud.frame.size() * 2)
        cloud.frame.set_top_left(origin)
        cloud.is_ephemeral = True
        world.children.append(cloud)
        return cloud

    def _setup_seeker(self):
        y_offset = self.cloud.frame.height() - self.target.frame.height()
        seeker = self.cloud.install(Seeker)
        seeker.follow(
            self.target,
            position=SeekerTargetPosition.ABOVE,
            offset=Point(0, y_offset),
            auto_adjust_speed=True
        )

    def _cleanup(self):
        if self.cloud in self.cloud.world.children:
            self.cloud.kill()
            self.cloud.world.children.remove(self.cloud)

    def _cloud_species(self):
        return Species(
            id="fantozzi",
            capabilities=[
                "AnimatedSprite",
                "AnimationsProvider",
                "LinearMovement",
                "PetsSpritesProvider"
            ],
            drag_path="front",
            movement_path="front",
            speed=2,
            z_index=200
        )